pkg/services/ngalert/api/api_provisioning.go

```diff
@@ -274,9 +278,12 @@ func (srv *ProvisioningSrv) RoutePostAlertRule(c *models.ReqContext, ar definiti
 }
 
 func (srv *ProvisioningSrv) RoutePutAlertRule(c *models.ReqContext, ar definitions.ProvisionedAlertRule, UID string) response.Response {
-	updated := ar.UpstreamModel()
+	updated, err := ar.UpstreamModel()
+	if err != nil {
+		ErrResp(http.StatusBadRequest, err, "")
+	}
 	updated.UID = UID
-	updatedAlertRule, err := srv.alertRules.UpdateAlertRule(c.Req.Context(), ar.UpstreamModel(), alerting_models.ProvenanceAPI)
+	updatedAlertRule, err := srv.alertRules.UpdateAlertRule(c.Req.Context(), updated, alerting_models.ProvenanceAPI)
```

Comment:
This is a side-effect of this PR, we used the UID from the payload, but now we use it from the path. `updated` was never used before, just the UID was set and then nothing would happen with it.

Reply:
Lol, good catch. I feel like a linter should be able to catch this sort of thing since Go does not support overriding field setters?

For example, If I do this, the compiler complains:
```
unusedVar := doSomethingThatReturnsStructNotPtr()
<never touch unusedVar again>
```
but then nothing catches if I do this:
```
unusedVar := doSomethingThatReturnsStructNotPtr()
unusedVar.Foo = Bar
<never touch unusedVar again>
```
but we still clearly didn't do anything with `unusedVar`. The assignment will never have a side effect for sure because Go doesn't support overriding setters and it's not a ptr type. So, in pretty much all cases the above could theoretically be flagged, right? :thinking:

Reply:
Cool, i'm not the only one: https://github.com/golang/go/issues/449